server/src/Main.hs

```diff
@@ -143,27 +199,118 @@ app origin rpIdHash db pending registryVar = do
+  -- FIXME
+  registry <- Scotty.liftAndCatchIO $ readTVarIO registryVar
+  now <- Scotty.liftAndCatchIO dateCurrent
+  result <- case WA.verifyRegistrationResponse origin rpIdHash registry now options cred of
+    Failure errs@(err :| _) -> do
+      Scotty.liftAndCatchIO $ TIO.putStrLn $ "Register complete had errors: " <> Text.pack (show errs)
+      fail $ show err
+    Success result -> pure result
+  Scotty.liftAndCatchIO $ TIO.putStrLn $ "Register complete result: " <> jsonText result
+  -- if the credential was succesfully attested, we will see if the
+  -- credential doesn't exist yet, and if it doesn't, insert it.
+  Scotty.liftAndCatchIO $
+    Database.withTransaction db $ \tx -> do
+      -- If a credential with this id existed already, it must belong to the
+      -- current user, otherwise it's an error. The spec allows removing the
+      -- credential from the old user instead, but we don't do that.
+      mexistingEntry <- Database.queryCredentialEntryByCredential tx (WA.ceCredentialId $ WA.rrEntry result)
+      case mexistingEntry of
+        Nothing -> do
+          Database.insertUser tx $ WA.corUser options
+          Database.insertCredentialEntry tx $ WA.rrEntry result
+          pure ()
+        Just existingEntry | userHandle == WA.ceUserHandle existingEntry -> pure ()
+        Just differentEntry -> do
+          TIO.putStrLn $ "Register complete credential already belongs to the user credential entry: " <> jsonText differentEntry
+          fail "This credential is already registered"
+  setAuthenticatedAs db userHandle
+  let result = String "success"
+  Scotty.liftAndCatchIO $ TIO.putStrLn $ "Register complete => " <> jsonText result
+  Scotty.json result
+
+-- | Starts the login procedure. In this function we receive the intent to
+-- login from the client, retrieve the userdata from the database, and reply
+-- with the
+-- [request options](https://www.w3.org/TR/webauthn-2/#dictdef-publickeycredentialrequestoptions).
+beginLogin :: Database.Connection -> PendingCeremonies -> Scotty.ActionM ()
 beginLogin db pending = do
+  -- Receive login name from the login field
   accountName <- WA.UserAccountName <$> Scotty.jsonData @Text
   Scotty.liftAndCatchIO $ TIO.putStrLn $ "Login begin <= " <> jsonText accountName
 
+  -- Retrieve account details from the database
   credentials <- Scotty.liftAndCatchIO $
     Database.withTransaction db $ \tx -> do
       Database.queryCredentialEntriesByUser tx accountName
   when (null credentials) $ do
     Scotty.liftAndCatchIO $ TIO.putStrLn "Login begin error: User not found"
     Scotty.raiseStatus HTTP.status404 "User not found"
+
+  -- Create credential options from the credential retrieved from the database
+  -- and insert the options into the pending ceremonies. This server stores the
+  -- entire options, but this isn't actually necessary for most RP
+  -- implementations. See the documentation of `WA.CredentialOptions` for more
```

Comment:
Wording it like this would indicate that it is needed for some RP implementations, which is not the case as far as I know.